src/main.rs

```diff
@@ -0,0 +1,295 @@
+use clap::{App, Arg, SubCommand};
+use serde::{Deserialize, Serialize};
+use std::error::Error;
+
+static PARENT_FOLDER: &str = ".graft.lock";
+static TMP_FOLDER: &str = "tmp";
+static CONFIG_FILE: &str = "graft.txt";
+static LOCK_FILE: &str = "conflict.lock";
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Repo {
+    upstream_path: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Config {
+    repos_to_fetch: Vec<Repo>,
+}
+
+fn read_config<P: AsRef<std::path::Path>>(path: P) -> Result<Config, Box<dyn Error>> {
+    let file = std::fs::File::open(path)?;
+    let reader = std::io::BufReader::new(file);
+    let config = serde_json::from_reader(reader)?;
+    Ok(config)
+}
+
+fn write_file_content(content: String, file_path: &str) -> Result<(), Box<dyn Error>> {
+    match std::path::Path::new(file_path).parent() {
+        Some(p) => {
+            std::fs::create_dir_all(p);
+        }
+        None => {}
+    }
+    let mut out = std::fs::File::create(file_path)
+        .expect(format!("ERROR: failed to create file: {:?}", file_path).as_str());
+    std::io::copy(&mut content.as_bytes(), &mut out).expect("ERROR: failed to copy content");
+    Ok(())
+}
+
+fn extract_files(path: &str) -> Result<(), Box<dyn Error>> {
+    let file = std::fs::File::open(path).unwrap();
+    let mut ar = tar::Archive::new(flate2::read::GzDecoder::new(file));
+    ar.unpack(format!("{}/{}", PARENT_FOLDER, TMP_FOLDER).as_str())
+        .unwrap();
+    Ok(())
+}
+
+fn download_file(url: &str) -> Result<String, Box<dyn Error>> {
+    let response = reqwest::blocking::get(url)?;
+    let fname = response
+        .url()
+        .path_segments()
+        .and_then(|segments| segments.last())
+        .and_then(|name| if name.is_empty() { None } else { Some(name) })
+        .unwrap_or("tmp.bin")
+        .to_owned();
+    let mut dest = std::fs::File::create(&fname)?;
+    let content = response
+        .bytes()
+        .expect(format!("ERROR: failed to download file from path: {:?}", url).as_str());
+    std::io::copy(&mut content.as_ref(), &mut dest)?;
+    Ok(fname)
```

Comment:
I was wondering how to make this testable. Lets write some test helpers, eg to see if we are running in test mode, and if so instead of downloading the file from internet, it gets it from some test case directory where we would checkin .tar.gz files, and also some directory diff function, to ensure proper directory structure is getting created.